### PR TITLE
minimum_st_node_cut breaks if source and target are connected

### LINKED ISSUE #1812: minimum_node_cut gives a wrong result if there is an edge between source and target nodes in an directed Graph
For the simple graph generated by the following code: 

    G = nx.Graph()
    G.add_nodes_from([0,1,2,3,4])
    G.add_edges_from([(0,1),(0,2),(1,3),(3,4),(2,3)], weight = 1)
    print nx.algorithms.connectivity.minimum_node_cut(G,1,0)

The output is `set([2])`.

While clearly, even with removing node 2 there would still be a path between 0 and 1 which is directly between them. I believe the proper output should be an empty set indicating there is no possible node cut.

### PR DESCRIPTION
**Pull request by Arafatk**
Fixes [ #1812](https://github.com/networkx/networkx/issues/1812)
For the function minimum_node_cut 
When source and sink are specified for minimum_node_cut we use the function   **minimum_st_node_cut**
which returns a set of nodes of minimum cardinality that disconnect source  from target in a graph.

For a given graph this is an example  of this function
correct case
Here the *blue vertices represent the output of the function minimum_node_cut.*
It can easily be observed that removing vertices 8,2,11 would disconnect 1,7. Hence the output is correct.

Now the **bug arises when the source and sink are adjacent**
For example 
bug case
Here the function wrongly outputs 2,8 as vertices which is incorrect. This happens because the algorithm  in following steps
1. Build Auxiliary Graph
2. Build residual network 
3. Get Cut Edges using Edmonds Karp 
4. Get cut nodes
This helps you get the minimum number of nodes between source to sink which could disconnect the source and sink but these vertices include the source and the sink themselves which cannot be excluded for obvious reasons. So in this case  we see that removing vertices 2,8 would remove all connections between 1,3 but an edge(1,3) still remains to connect them. 

The solution to this bug is to output an empty set in case the source is connected to the sink or the sink is connected to source. 
For Example:- 
correction

On applying the fix the following output is obtained. Indicating that it is impossible to delete a vertex that would disconnect 1,3.
@MridulS  Please suggest some ideas so that this fix could be added. The Travis build is failing here,because the doctests and other tests included in nosetests are not in conformity with these changes.

### CODE REVIEW

**Comment by dschult** [comment]
I would suggest adding a test to tests/test_cuts.py that checks for correct output in the problematic case.

While you are there, change the test at line 244 to use ```G = nx.path_graph(5)``` and add a comment something like ```# auxiliary should be a mapping not a graph```.  This will fix the travis-CI fails and you can move your new code back to the top of the function (don't want to compute the auxiliary if we don't need it.)

**Comment by Arafatk** [comment]
@dschult As per your advice, I edited the existing  tests, and added new tests that checks for correct output in the problematic case.

File: networkx/algorithms/connectivity/tests/test_cuts.py [hunk 1]
```diff
@@ -6,7 +6,7 @@
 from networkx.algorithms.flow import (edmonds_karp, preflow_push,
                                       shortest_augmenting_path)
 from networkx.utils import arbitrary_element
-
+from networkx.algorithms.connectivity import minimum_st_edge_cut,minimum_node_cut
```
Comment (dschult) [inline networkx/algorithms/connectivity/tests/test_cuts.py]: put this back to a blank line as the import is done already 3-4 lines up.

**Comment by dschult** [comment]
I might be wrong on the definition here, but I think the edge_cut routines should not change... If two nodes are directly connected then that edge must be in the edge_cut. And the edge_cut does exist. The problem arises for node_cuts because you can't remove the source and target nodes, so an edge between them can't be cut.

So I think you'll have to undo some of your changes.

**Comment by Arafatk** [comment]
@dschult You are right,edge_cut exists even if an edge exists between source and node and hence it must not be changed. I removed unnecessary changes and kept the test case for minimum_st_node_cut.

**Comment by dschult** [comment]
OK... now we're getting to the silly details at the end...
Could you adjust spacing in the new code to conform to PEP8 standards and then "squash your commits". 

It looks like spacing is the only PEP8 issue... need spaces around ```=```, ```==``` and after each comma. So ```G.has_edge(u,v)``` becomes ```G.has_edge(u, v)```. Lots of commas...:) If that ends up making the line longer than 80 characters you should split the line onto the next line, or create a variable to hold an intermediate result, or in your case, simplifying the example by removing an edge. 

To [squash your commits](https://github.com/ginatrapani/todo.txt-android/wiki/Squash-All-Commits-Related-to-a-Single-Issue-into-a-Single-Commit), use ```git rebase -i HEAD~6```, change the ```pick``` commands for all commits after the first so they are squashed to the commit above them, save, edit the resulting commit messages as appropriate and save again. In your commit messages the first line's first 50 chars become the title of the commit. Skip a line and then put any details that you want in the description of the commit.  There may be better help on squashing if you search for it.

**Comment by jfinkels** [comment]
Here is GitHub's help on rebasing: https://help.github.com/articles/about-git-rebase/

**Comment by Arafatk** [comment]
@dschult  Thanks a lot for the valuable pointers. I have handled the PEP8  guidelines but their are a lot of issues with rebasing because of my college network proxy(I perform all the changes on git  website itself and not from terminal). Is it okay,if i close this pr and make a new one with the exact same changes?

**Comment by MridulS** [comment]
Git can be confusing at first. Don't worry :) Send in a new PR.
On Jan 7, 2016 9:53 PM, "Arafat" <notifications@github.com> wrote:

@dschult <https://github.com/dschult> Thanks a lot for the valuable
pointers. I have handled the PEP8 guidelines but their are a lot of issues
with rebasing because of my college network proxy(I perform all the changes
on git website itself and not from terminal). Is it okay,if i close this pr
and make a new one with the exact same changes?

—
Reply to this email directly or view it on GitHub
<https://github.com/networkx/networkx/pull/1902#issuecomment-169714469>.

**Comment by OrkoHunter** [comment]
@Arafatk I see you are using the GitHub interfaces, which is not recommended in the most of the cases. :) Regarding the proxy issue, you'll need to just stick with the `https` protocol because you will not be able to get `ssh` and `git` working easily. Ping me if you need more help.

**Comment by dschult** [comment]
I think you should be able to use git via the ```https``` remote addresses. If you can see it on a browser you should be able to get git to interact with it. It'll be worth spending some time getting this to work. I don't have a lot of experience with it but @OrkoHunter can help. The theory is to ```git clone``` and ```git remote add``` using the https://github.com/networkx/networkx.git addresses.

**Pull request closed without merging**